File: src/main/java/com/crazicrafter1/lootcrates/PlayerLog.java
package com.crazicrafter1.lootcrates;

import org.bukkit.command.CommandSender;
import org.bukkit.configuration.file.YamlConfiguration;

import javax.annotation.Nonnull;
import java.io.File;
import java.text.SimpleDateFormat;
import java.util.*;

public class PlayerLog {

    private static final File PATH = new File(LCMain.get().getDataFolder(), "players");

    //public boolean antiExploit = false;
    private static final HashMap<UUID, PlayerLog> LOGS = new HashMap<>();
    // get crate open information and other stuff to analyze possible duping or whatever
    //
    private final HashMap<String, List<String>> crates = new HashMap<>();

    public void increment(String id) {
        List<String> dateList = crates.computeIfAbsent(id, k -> new ArrayList<>());
        SimpleDateFormat formatter = new SimpleDateFormat("yyyy-MM-dd 'at' HH:mm:ss z");
        Date date = new Date(System.currentTimeMillis());
        dateList.add(formatter.format(date));
    }


    @Nonnull
    public static PlayerLog get(UUID uuid) {
        // if not present add

        PlayerLog stat = PlayerLog.LOGS.get(uuid);
        if (stat == null) {
            stat = new PlayerLog();
            PlayerLog.LOGS.put(uuid, stat);
        }

        return stat;
    }

    public static void saveAll(@Nonnull CommandSender sender) {
        try {
            for (Map.Entry<UUID, PlayerLog> entry : PlayerLog.LOGS.entrySet()) {
                String rawUUID = entry.getKey().toString();
                YamlConfiguration playerConfig = new YamlConfiguration();
                for (Map.Entry<String, List<String>> entry1 : entry.getValue().crates.entrySet()) {
                    playerConfig.set(entry1.getKey(), entry1.getValue());
                }
                playerConfig.save(new File(PATH, rawUUID + ".yml"));
            }
        } catch (Exception e) {
            LCMain.get().notifier.severe(sender, String.format(Lang.CONFIG_ERROR10, e.getMessage()));
        }
    }

    public static void loadAll(@Nonnull CommandSender sender) {
        try {
            if (LCMain.get().rev < 5)
                return;

            if (!PATH.exists() || !PATH.isDirectory())
                return;

            // playerStatsPath.mkdirs();

            File[] files = PATH.listFiles(); // TODO use Paths.walk

            for (File file : files) {
                try {
                    String rawUUID = file.getName().replace(".yml", "");
                    UUID uuid = UUID.fromString(rawUUID);

                    YamlConfiguration playerConfig = new YamlConfiguration();
                    playerConfig.load(file);

                    PlayerLog stat = new PlayerLog();
                    PlayerLog.LOGS.put(uuid, stat);
                    for (String id : playerConfig.getKeys(false))
                        stat.crates.put(id, playerConfig.getStringList(id));

                } catch (Exception ignored) {}
            }
        } catch (Exception e) {
            LCMain.get().notifier.severe(sender, String.format(Lang.CONFIG_ERROR11, e.getMessage()));
        }
    }

}

File: src/main/java/com/crazicrafter1/lootcrates/PlayerLog.kt
package com.crazicrafter1.lootcrates

import org.bukkit.command.CommandSender
import org.bukkit.configuration.file.YamlConfiguration
import java.io.File
import java.text.SimpleDateFormat
import java.util.*

class PlayerLog {
    // get crate open information and other stuff to analyze possible duping or whatever
    //
    private val crates = HashMap<String?, MutableList<String>>()
    fun increment(id: String?) {
        val dateList = crates.computeIfAbsent(id) { k: String? -> ArrayList() }
        val formatter = SimpleDateFormat("yyyy-MM-dd 'at' HH:mm:ss z")
        val date = Date(System.currentTimeMillis())
        dateList.add(formatter.format(date))
    }

    companion object {
        private val PATH: File = File(LCMain.Companion.get()!!.getDataFolder(), "players")

        //public boolean antiExploit = false;
        private val LOGS = HashMap<UUID, PlayerLog>()

        @Nonnull
        operator fun get(uuid: UUID): PlayerLog {
            // if not present add
            var stat = LOGS[uuid]
            if (stat == null) {
                stat = PlayerLog()
                LOGS[uuid] = stat
            }
            return stat
        }

        fun saveAll(@Nonnull sender: CommandSender?) {
            try {
                for ((key, value) in LOGS) {
                    val rawUUID = key.toString()
                    val playerConfig = YamlConfiguration()
                    for ((key1, value1) in value.crates) {
                        playerConfig[key1!!] = value1
                    }
                    playerConfig.save(File(PATH, "$rawUUID.yml"))
                }
            } catch (e: Exception) {
                LCMain.Companion.get()!!.notifier!!.severe(sender!!, String.format(Lang.MESSAGE_STATS_ERROR10, e.message))
            }
        }

        fun loadAll(@Nonnull sender: CommandSender?) {
            try {
                if (LCMain.Companion.get()!!.rev < 5) return
                if (!PATH.exists() || !PATH.isDirectory()) return

                // playerStatsPath.mkdirs();
                val files = PATH.listFiles() // TODO use Paths.walk
                for (file in files) {
                    try {
                        val rawUUID = file.getName().replace(".yml", "")
                        val uuid = UUID.fromString(rawUUID)
                        val playerConfig = YamlConfiguration()
                        playerConfig.load(file)
                        val stat = PlayerLog()
                        LOGS[uuid] = stat
                        for (id in playerConfig.getKeys(false)) stat.crates[id] = playerConfig.getStringList(id!!)
                    } catch (ignored: Exception) {
                    }
                }
            } catch (e: Exception) {
                LCMain.Companion.get()!!.notifier!!.severe(sender!!, String.format(Lang.MESSAGE_STATS_ERROR11, e.message))
            }
        }
    }
}
